File: src/ForzaStudio/ForzaTrackSection.cs
using Microsoft.Xna.Framework;

namespace ForzaStudio;

public class ForzaTrackSection
{
	public ForzaTrackBin Parent;

	public EndianStream Stream;

	public ForzaTrackSubSection[] SubSections;

	public string Name;

	public string Type;

	public Vector3 Offset;

	public ForzaTrackSection(ForzaTrackBin parent)
	{
		Parent = parent;
		Stream = parent.Stream;
		Utilities.AssertEquals(Stream.ReadUInt32(), 1u);
		new Vector3(Stream.ReadSingle(), Stream.ReadSingle(), Stream.ReadSingle());
		Utilities.AssertEquals(Stream.ReadSingle(), 0f);
		new Vector3(Stream.ReadSingle(), Stream.ReadSingle(), Stream.ReadSingle());
		Utilities.AssertEquals(Stream.ReadSingle(), 0f);
		new Vector3(Stream.ReadSingle(), Stream.ReadSingle(), Stream.ReadSingle());
		Utilities.AssertEquals(Stream.ReadSingle(), 0f);
		Utilities.AssertEquals(Stream.ReadUInt32(), 1u);
		Name = Stream.ReadASCII(Stream.ReadInt32()).ToLowerInvariant();
		string[] array = Name.Split('_');
		Type = array[0].ToLowerInvariant();
		Name = Name.Substring(Name.IndexOf("_") + 1);
		Utilities.AssertEquals(Stream.ReadUInt32(), 2u);
		uint num = Stream.ReadUInt32();
		uint size = Stream.ReadUInt32();
		ForzaVertex[] array2 = new ForzaVertex[num];
		for (int i = 0; i < num; i++)
		{
			ref ForzaVertex reference = ref array2[i];
			reference = new ForzaVertex(parent.Version, ForzaVertexType.Track, Stream, size);
		}
		Utilities.AssertEquals(Stream.ReadUInt32(), 1u);
		BoundingBox.FromVertices(array2);
		SubSections = new ForzaTrackSubSection[Stream.ReadUInt32()];
		for (int j = 0; j < SubSections.Length; j++)
		{
			ForzaTrackSubSection forzaTrackSubSection = new ForzaTrackSubSection(this);
			forzaTrackSubSection.Vertices = Utilities.GenerateVertices(array2, ref forzaTrackSubSection.Indices);
			for (int k = 0; k < forzaTrackSubSection.Vertices.Length; k++)
			{
				forzaTrackSubSection.Vertices[k].texture0 *= forzaTrackSubSection.UVtile;
				forzaTrackSubSection.Vertices[k].texture1 *= forzaTrackSubSection.UVtile;
				forzaTrackSubSection.Vertices[k].texture0 += forzaTrackSubSection.UVoffset;
				forzaTrackSubSection.Vertices[k].texture1 += forzaTrackSubSection.UVoffset;
				forzaTrackSubSection.Vertices[k].texture0.Y = 1f - forzaTrackSubSection.Vertices[k].texture0.Y;
				forzaTrackSubSection.Vertices[k].texture1.Y = 1f - forzaTrackSubSection.Vertices[k].texture1.Y;
			}
			if (forzaTrackSubSection.IndexType == IndexType.TriStrip)
			{
				forzaTrackSubSection.Indices = Utilities.GenerateTriangleList(forzaTrackSubSection.Indices, forzaTrackSubSection.FaceCount);
			}
			SubSections[j] = forzaTrackSubSection;
		}
	}
}

File: src/py/models/forza_track_section.py
from typing import List
from utils.forza_version import ForzaVersion

class ForzaTrackSection:
    def __init__(self, forza_version: ForzaVersion):
        self.SubSections: List["ForzaTrackSubSection"] = []
        self.Name: str = ""
        self.Type: str = ""
        self.Offset: "Vector3" = Vector3(0.0, 0.0, 0.0)

        # Header / transforms (pattern of 1, vec3, 0.0, vec3, 0.0, vec3, 0.0, 1)
        Utilities.assert_equals(self.Stream.read_uint32(), 1)
        _ = Vector3(self.Stream.read_single(), self.Stream.read_single(), self.Stream.read_single())
        Utilities.assert_equals(self.Stream.read_single(), 0.0)
        _ = Vector3(self.Stream.read_single(), self.Stream.read_single(), self.Stream.read_single())
        Utilities.assert_equals(self.Stream.read_single(), 0.0)
        _ = Vector3(self.Stream.read_single(), self.Stream.read_single(), self.Stream.read_single())
        Utilities.assert_equals(self.Stream.read_single(), 0.0)
        Utilities.assert_equals(self.Stream.read_uint32(), 1)

        # Name/type
        name_len = self.Stream.read_int32()
        self.Name = self.Stream.read_ascii(name_len).lower()
        parts = self.Name.split('_')
        self.Type = parts[0].lower()
        self.Name = self.Name[self.Name.index('_') + 1:]

        # Vertex block
        Utilities.assert_equals(self.Stream.read_uint32(), 2)
        num = self.Stream.read_uint32()
        size = self.Stream.read_uint32()

        base_vertices: List["ForzaVertex"] = [None] * num  # type: ignore
        for i in range(num):
            base_vertices[i] = ForzaVertex(forza_version, ForzaVertexType.Track, self.Stream, size)

        Utilities.assert_equals(self.Stream.read_uint32(), 1)
        BoundingBox.from_vertices(base_vertices)

        # Subsections
        sub_count = self.Stream.read_uint32()
        self.SubSections = [None] * sub_count  # type: ignore

        for j in range(sub_count):
            sub = ForzaTrackSubSection(self)

            # Generate per-subsection vertices (and possibly remapped indices)
            # If your Utilities.generate_vertices only returns vertices, drop the second assignment.
            verts, indices = Utilities.generate_vertices(base_vertices, sub.Indices)
            sub.Vertices = verts
            sub.Indices = indices

            # UV adjustments
            for v in sub.Vertices:
                v.texture0 *= sub.UVtile
                v.texture1 *= sub.UVtile
                v.texture0 += sub.UVoffset
                v.texture1 += sub.UVoffset
                v.texture0.y = 1.0 - v.texture0.y
                v.texture1.y = 1.0 - v.texture1.y

            # Convert tristrips to triangle list if needed
            if sub.IndexType == IndexType.TriStrip:
                sub.Indices = Utilities.generate_triangle_list(sub.Indices, sub.FaceCount)

            self.SubSections[j] = sub
